### PR TITLE
Harmonize watchers endpoints

### PR DESCRIPTION
**Pull request by NobodysNightmare**
:warning: Please merge this into `release/4.3` :warning: 
## OpenProject work package

https://community.openproject.org/work_packages/17755
## Description

This PR harmonizes the documentation around the watchers endpoint and other similar endpoints.
It also applies the relevant changes to the implementation and aligns the implementation with the specification.

### CODE REVIEW

**Comment by ulferts** [comment]
This got me quite confused so please ask if my writing is not understandable:

It used to be my understanding that the `watch` link on work package was designed to add oneself as the watcher of the work package. Apparently, that understanding was wrong - there seems to have been the possibility to add arbitrary users at this endpoint. At least that what it says in the docu. The code says differently. But the new code at least allows for adding arbitrary watchers via the endpoint. 

What do we need the `addWatcher` link for then? Is it to signal, that the user can add arbitrary users instead of only oneself? The addWatcher link is described wrongly as it has the href `"/api/v3/work_packages/1528/watchers{?user_id}"` and states to be templated. The code however does not support providing the user_id as a parameter. Only payload is allowed. 

The permissions checked for the `watch` link read as if that link is only shown if it should be used for adding oneself. The `add_work_package_watchers` permission is not checked there. It is checked for the `addWatcher` link but that link does not take into account whether the user has the `view_work_packages` permission which should suffice for adding oneself.

The `addWatcher` link could probably be removed if it weren't for the need to signal to the front end whether the user is allowed to add watchers other than oneself. This could of course be mitigated by the front-end fetching the list of available_watchers and noting that he is the only one on that list but it would be awkward and error prone (there might be only the one user in the project for now).

On top of this confusion, and the links being displayed when they should not, there is a bug in combination with the front end. When I click on the star icon in the work package split view I can no longer watch the work package. That again is due to the user_id parameter.

**Comment by floriank** [comment]
this could either use a rebase and some love or a close.

**Comment by NobodysNightmare** [comment]
It needs a merge and love, **not** a close ^^

I will do that _soon_.

**Comment by Azure7111** [comment]
The specs seem to fail legitimately or at least somewhat consistently.

File: doc/apiv3-documentation.apib [hunk 1]
```diff
@@ -114,12 +115,17 @@ For example a work package without an assignee will still have an assignee link,
...
-Templated links are links that contain client replacable parts. Replaceable parts are enclosed in braces. For example
+Templated links are links that contain client replaceable parts. Replaceable parts are enclosed in braces. For example
 the link `api/v3/example/{id}` is not complete in itself, but the client needs to replace the string `{id}` itself.
 As of now the API does not indicate the valid replacement values.
 
 The `method` indicates which HTTP verb the client *must* use when following the link for the intended purpose.
 
+If a `payload` is specified, it needs to be sent as the body of the request to achieve the desired result (e.g. perform the
+action represented by the link). If no `payload` is specified, there is either no payload to send or there are multiple valid
+payloads. A payload might also be templated slightly. If the `templated` property is true, a payload might contain replaceable parts
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib]: "...templated _slightly_"? I'd say either something is templated or it isn't.

File: doc/apiv3-documentation.apib [hunk 2]
```diff
@@ -3172,21 +3180,21 @@ the human readable name of custom fields.*
...
                         "method": "post",
-                        "data": {
-                            "user_id": 1
-                        },
-                        "title": "Watch work package"
+                        "payload": {
+                            "user": { "href": "/api/v3/users/1" }
+                        }
                     },
                     "addWatcher": {
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3193]: What about "removeWatcher"? Since the ticket does not explicitly state the necessity of either one I'd expected it for the sake of consistency. Also: Should both actions go to the same endpoint but use different http verbs (like POST and DELETE) or rather use separate endpoints?

File: doc/apiv3-documentation.apib [hunk 3]
```diff
@@ -3390,41 +3430,306 @@ For more details and all possible responses see the general specification of [Fo
 
     [Example Form][]
 
-## Add Watcher [/api/v3/work_packages/{work_package_id}/watchers]
++ Response 403 (application/hal+json)
 
-## Add watcher [POST]
+    Returned if the work package does not exist or the client does not have sufficient permissions to see it.
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib]: Copy pasta? :spaghetti:  Here we should already have the permission to see a work package but lack the permission to edit it.

File: doc/apiv3-documentation.apib (hunk 1 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: It is not fully templatable, I only added semantics for templating inside strings in the payload.

If you know a better word, tell me...

Comment (Azure7111) [inline doc/apiv3-documentation.apib]: Hmm, I don't care that much, but how about "parameters"?

File: doc/apiv3-documentation.apib (hunk 3 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: mmmhhhhh, pasta....

:spaghetti: :spaghetti: :spaghetti: :spaghetti:

File: doc/apiv3-documentation.apib [hunk 4]
```diff
@@ -3390,41 +3430,306 @@ For more details and all possible responses see the general specification of [Fo
...
+                            "self": {
+                                "href": "/api/v3/users/1",
+                                "title": "John Sheppard - j.sheppard"
+                            },
+                            "lock": {
+                                "href": "/api/v3/users/1/lock",
+                                "title": "Set lock on j.sheppard"
+                                "method": "POST"
+                            },
+                            "delete": {
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3495]: I know this is only a slight thing and probably just my own insecurity speaking. I also think this would take more work than might be appropriate, but: As I read this link my first thought was: "And this is probably a link to remove this user as watcher from this WP" because I am looking at a collection of watchers. Immerdiately thereafter I realized that this is wrong, e.g. because it makes no sense to "lock" a watcher. But: What do you think about postfixing the link names with "User"? (i.e. "lockUser" and "deleteUser")

I only wrote about this point because my misunderstanding might have made me delete a user, whereas I only wanted to remove a watcher.

File: doc/apiv3-documentation.apib (hunk 2 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3193]: sure :cat2:

File: doc/apiv3-documentation.apib (hunk 4 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3495]: I see your point.
I would argue to not address it now, because those link names make perfect sense on a user object alone and are also already in use.

What might make sense in a future revision of the API: Returning watcher objects instead of user objects (e.g. `api/v3/watchers/13`). Database side watchers have a unique id (e.g. if you watch two work packages, there are two watcher objects for your user account) and might make it more clear _what_ you are deleting.

However, the initial thought of not exposing watcher objects was to provide an abstraction from the database concept and call everything a user. This is now one caveat.

**tl;dr:** For now I would leave it as-is, but if it turns out that there will be more confusion that gains, we might want to reverse course...

Comment (Azure7111) [inline doc/apiv3-documentation.apib:3495]: Okay as we agreed: We will not do anything about this until the amount of inadvertently deleted users becomes too damn high.

File: doc/apiv3-documentation.apib [hunk 5]
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
...
             }
 
+## Add watcher [POST]
+
+Adds a watcher to the specified work package.
+
+The request is expected to contain a single JSON object, that contains a link object under the `user` key.
+
+The response will be user added as watcher.
+In case the user was already watching the work package an `HTTP 200` is returned, an
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3590]: Have you considered to place this remark into the section of the corresponding response?

Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3590]: I remember something like the blueprint not working when I do that...

edit: [found it](https://github.com/opf/openproject/commit/89ef5f648fe264ef82358405b69dd030d28ce1c4)

File: doc/apiv3-documentation.apib [hunk 6]
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
...
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:InvalidRequestBody",
+                "message": "The request body was not a single JSON object."
+            }
+
++ Response 403 (application/hal+json)
+
+    Returned if the client does not have sufficient permissions.
+
+    **Required permission:**
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib]: It's more than one ;)

File: doc/apiv3-documentation.apib [hunk 7]
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
...
+
+    + Body
+
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:ResourceTypeMismatch",
+                "message": "Expected resource of type 'User', but got a 'Status'.",
+            }
+
 ## Remove Watcher [/api/v3/work_packages/{work_package_id}/watchers/{id}]
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3679]: I know you did not touch this but what about renaming id -> user_id?

File: doc/apiv3-documentation.apib [hunk 8]
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
...
+
++ Response 404 (application/hal+json)
+
+    Returned in one of two cases:
+
+    Either the work package does not exist or the client does not have sufficient permissions to see it.
+
+    **Required permission:** view work package
+
+    Or the specified user does not exist at all.
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3719]: That would be a third case, you might to consider rephrasing your sentence above which states that there are only two cases. Also: Maybe pull this sentence, or its statement for that matter, above the line with Required permission.

File: doc/apiv3-documentation.apib [hunk 9]
```diff
@@ -3390,41 +3436,306 @@ For more details and all possible responses see the general specification of [Fo
...
+
+    + Body
+
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "The specified work package does not exist."
+            }
+
+## Available Watchers [/api/v3/work_packages/{id}/available_watchers]
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3731]: It appears to me as if this is a more general inconsistency in our API documentation: Somewhere above you went the extra mile and called this `id` actually `work_package_id`.

File: doc/apiv3-documentation.apib [hunk 10]
```diff
@@ -3492,10 +3803,10 @@ For more details and all possible responses see the general specification of [Fo
 
 ## Available watchers [GET]
 
-Gets a list of users that can watch the work package.
+Gets a list of users that are able to be watchers of the specified work package.
 
 + Parameters
-    + work_package_id (required, integer, `1`) ... Work package id
+    + id (required, integer, `1`) ... work package id
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3809]: Could you enlighten me with the rationale for this change?

File: doc/apiv3-documentation.apib [hunk 11]
```diff
@@ -3492,10 +3803,10 @@ For more details and all possible responses see the general specification of [Fo
 
 ## Available watchers [GET]
 
-Gets a list of users that can watch the work package.
+Gets a list of users that are able to be watchers of the specified work package.
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib:3806]: This is totally fine with me, but what is the gain here?

File: doc/apiv3-documentation.apib [hunk 12]
```diff
@@ -3507,6 +3818,8 @@ Gets a list of users that can watch the work package.
 
     **Required permission:** add work package watchers
 
+    *Note that you will only receive this error, if you are at least allowed to see the corresponding project.*
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib]: I bet this is :spaghetti: Copy pasta again. This should probably read "work package" instead of "project".

File: doc/apiv3-documentation.apib [hunk 13]
```diff
@@ -3515,6 +3828,20 @@ Gets a list of users that can watch the work package.
...
+    Returned if the project does not exist or the client does not have sufficient permissions to see it.
+
+    **Required permission:** view project
+
+    + Body
+
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "The specified project does not exist."
```
Comment (Azure7111) [inline doc/apiv3-documentation.apib]: This seems weird, I'd have expected this entire response to revolve around the specified _work package_.

File: lib/api/decorators/link_object.rb [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      end
+
+      property :href,
+               exec_context: :decorator,
+               getter: -> (*) {
+                 id = represented.send(@getter) if represented
+
+                 return nil unless id
+
+                 api_v3_paths.send(@path, id)
```
Comment (Azure7111) [inline lib/api/decorators/link_object.rb:55]: Would this work for paths which do not require an id?

File: doc/apiv3-documentation.apib (hunk 7 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3679]: I guess there will be more occurences of such errors.

One (anyone) could do this in a ticketless cleanup PR IMO, I am too lazy to do it in this PR TBH.

File: doc/apiv3-documentation.apib (hunk 9 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3731]: It is usually the follwoing rule:
- `id` if it is onyl one
- `specific_id` for the extra thing needed, in case it is more than one

File: doc/apiv3-documentation.apib (hunk 10 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3809]: As stated above:

Use the shortest possible name and only prepend words if that is neccessary.

If one wanted to change that, I would propose a separate PR that changes it consistently across the board.

File: doc/apiv3-documentation.apib (hunk 11 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:3806]: No idea...

edit: probably history of changes... or any other thought that crossed my mind...

File: doc/apiv3-documentation.apib (hunk 12 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: No copy pasta, but change pasta.
This endpoint was moved from project to work package and the message was not updated.

File: doc/apiv3-documentation.apib (hunk 13 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: the same copy pasta you noticed above... not sure why this comment did not get outdated...

File: lib/api/decorators/link_object.rb (hunk 14 above)
Comment (NobodysNightmare) [inline lib/api/decorators/link_object.rb:55]: no

File: lib/api/v3/work_packages/work_package_representer.rb [hunk 15]
```diff
@@ -301,13 +309,18 @@ def activities
...
-                                                  work_package: represented,
-                                                  current_user: current_user)
-          end
+          # TODO/LEGACY: why do we need to ensure a specific order here?
+          watchers = represented.watcher_users.order(User::USER_FORMATS_STRUCTURE[Setting.user_format])
+          total = watchers.count
+          self_link = api_v3_paths.work_package_watchers(represented.id)
+
+          # FIXME/LEGACY: we pass the WP as context?!? that makes a difference!!!
+          # tl;dr: the embedded user representer must not be better than any other user representer
```
Comment (Azure7111) [inline lib/api/v3/work_packages/work_package_representer.rb:318]: That sounds indeed odd. But have you found the reason why?

**Comment by ulferts** [comment]
@NobodysNightmare we should no longer merge this into 4.2. So once this is supposed to be done, please reissue against release/4.3

File: lib/api/v3/work_packages/work_packages_shared_helpers.rb [hunk 16]
```diff
@@ -38,26 +38,21 @@ def request_body
...
-              # we need to merge the JSON two times:
-              # In Pass 1 the representer only has custom fields for the current WP type
-              # After Pass 1 the correct type information is merged into the WP
-              # In Pass 2 the representer is created with the new type info and will be able
-              # to also parse custom fields successfully
-              merge_json_into_work_package!(work_package, request_body.to_json)
-              merge_json_into_work_package!(work_package, request_body.to_json)
-            rescue ::API::Errors::Form::InvalidResourceLink => e
-              fail ::API::Errors::Validation.new(e.message)
-            end
```
Comment (Azure7111) [inline lib/api/v3/work_packages/work_packages_shared_helpers.rb:60]: Could you shed some light on this removal?

File: lib/api/v3/work_packages/work_package_representer.rb (hunk 15 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/work_package_representer.rb:318]: The user representer actually makes use of the pased work package and provides more links in that case (watcher specific links).

This is in contrast to everything we discussed today about separating between user and watcher. However, I found it out of scope of this (already grown) PR to also address this issue, because at least our angular client relies on those fields ATM.

File: lib/api/v3/work_packages/work_packages_shared_helpers.rb (hunk 16 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/work_packages_shared_helpers.rb:60]: Sure:

There was a special case for one possible exception that could be thrown from the merge-method. Just to wrap this specific exception in another exception.

However, the cleaner approach is to immediately use the original exception here, as (per specification of this PR) there should be a separate error identifier for the given class of errors.

Comment (NobodysNightmare) [inline lib/api/v3/work_packages/work_packages_shared_helpers.rb:60]: Thinking about it now: This might be a bad thing because now that error will not cause a `Validation` error. Which in case is checked by [some other code](https://github.com/opf/openproject/blob/dev/lib/api/v3/work_packages/work_packages_shared_helpers.rb#L89).

I guess I will come up with a solution for that in my other PR: https://github.com/opf/openproject/pull/3166

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb [hunk 17]
```diff
@@ -493,7 +483,15 @@
...
-          expect(subject).to have_json_path('_links/addWatcher/href')
+          expect(subject).to be_json_eql(
+                               api_v3_paths.work_package_watchers(work_package.id).to_json)
+            .at_path('_links/addWatcher/href')
+        end
+
+        it 'should have a link to remove watcher' do
+          expect(subject).to be_json_eql(
+                               api_v3_paths.watcher('{user_id}', work_package.id).to_json)
+                               .at_path('_links/removeWatcher/href')
```
Comment (Azure7111) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb]: This indentation looks weird 0o. I'd expected these lines to be either more to the right, which might not be possible due to the line limit, or to the left (indented only once).

But since @houndci didn't have to say anything about this I'm a little confused and this is not important at all.

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb [hunk 18]
```diff
@@ -507,6 +505,39 @@
         end
       end
 
+      context 'when the user does not have the permission to remove watchers' do
+        before do
+          role.permissions.delete(:delete_work_package_watchers) and role.save
+        end
+
+        it 'should not have a link to add watcher' do
```
Comment (Azure7111) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb]: The test seems fine, its name not so much ;)

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb [hunk 19]
```diff
@@ -507,6 +505,39 @@
...
+          expect(subject).not_to have_json_path('_links/removeWatcher/href')
+        end
+      end
+
+      describe 'watchers link' do
+        context 'when the user is allowed to see watchers' do
+          it_behaves_like 'has an untitled link' do
```
Comment (Azure7111) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb:520]: I know that that the title of a link is optional, but is there a specific reason why we want to ensure that this one has no title?

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb [hunk 20]
```diff
@@ -507,6 +505,39 @@
...
+            let(:link) { 'watchers' }
+            let(:href) { api_v3_paths.work_package_watchers work_package.id }
+          end
+
+          it 'embeds the watchers as collection' do
+            is_expected.to be_json_eql('Collection'.to_json).at_path('_embedded/watchers/_type')
```
Comment (Azure7111) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb:526]: At this point I could imagine having a shared example checking the integrity of a collection. I'm not sure if we have something like that already. What do you think about that?

File: spec/lib/api/v3/work_packages/work_packages_shared_helpers_spec.rb [hunk 21]
```diff
@@ -67,19 +67,23 @@ def status(_code)
...
     end
 
     context 'invalid parameters' do
       context 'validation errors' do
-        let(:env) { { 'api.request.body' => { subject: '' } } }
+        let(:env) { { 'api.request.body' => { 'subject' => '' } } }
 
         it 'does not raise for validation errors' do
-          expect(subject.validation_errors.any?).to be_truthy
+          expect(subject.validation_errors).not_to be_empty
```
Comment (Azure7111) [inline spec/lib/api/v3/work_packages/work_packages_shared_helpers_spec.rb:81]: Nice catch! :+1:

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb (hunk 19 above)
Comment (NobodysNightmare) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb:520]: Specific tests are good.
I did not implement a title so I expect none.

If someone now decides to add a shitty title he will at least have to add a shitty spec too...

File: spec/lib/api/v3/work_packages/work_package_representer_spec.rb (hunk 20 above)
Comment (NobodysNightmare) [inline spec/lib/api/v3/work_packages/work_package_representer_spec.rb:526]: We have that already, however I don't want to test the representer for user collection here. I only want to verify that the wp representer utilizes such a representer and then there is another test checking whether the user collection representer does sane things.

Now I only need an approximation for "the wp representer uses a user collection representer in apropriate places". The approximation I've chosen here (and I think in other places too) is that there should be the type `Collection` at a specific path.

An evil developer could make this test green by re-implementing the user collection representer inline in the wp representer. But evil developers are hard to detect via unit tests ;-)

File: spec/requests/api/v3/watcher_resource_spec.rb [hunk 22]
```diff
@@ -182,47 +226,21 @@
...
-    end
-
-    it 'has users fit for watching embedded' do
-      expect(subject.body).to have_json_size(2).at_path('_embedded/elements')
-
-      watcher_hashes = parse_json(subject.body)['_embedded']['elements']
-      watcher_ids = watcher_hashes.map { |watcher_hash| watcher_hash['id'] }
-
-      expect(watcher_ids).to match_array([current_user.id, available_watcher.id])
+    it 'includes a user eligable for watching' do
```
Comment (Azure7111) [inline spec/requests/api/v3/watcher_resource_spec.rb]: typo

**Comment by NobodysNightmare** [comment]
CI is as green as it will get ^^

File: spec/lib/api/decorators/link_object_spec.rb [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+      context 'wrong namespace' do
+        let(:parsed_hash) {
+          {
+            'href' => '/api/v3/foos/42'
+          }
+        }
+
+        it 'throws an error' do
+          expect { representer.from_hash parsed_hash }.to raise_error(
+                                                            ::API::Errors::InvalidResourceLink)
```
Comment (NobodysNightmare) [inline spec/lib/api/decorators/link_object_spec.rb:123]: > 1 file inspected, no offenses detected

@houndci is a liar!

**Comment by Azure7111** [comment]
This PR has been merged into release/4.3, as seen in b3a8e0187ec4aad998dc66d669abb52531e66166.
